src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/CodeBasedDependencyAlgorithm.cs

```diff
@@ -118,23 +120,82 @@ public static void AddDependenciesDueToMethodCodePresence(ref DependencyList dep
-                dependencies.Add(factory.MethodEntrypoint(factory.InteropStubManager.GetClosedDelegateMarshallingStub(parameter)), "Delegate Marshalling Stub");
-                dependencies.Add(factory.MethodEntrypoint(factory.InteropStubManager.GetForwardDelegateCreationStub(parameter)), "Delegate Marshalling Stub");
+            if (metadataType != null)
+            {
+                if (metadataType.Name == "Marshal" && metadataType.Namespace == "System.Runtime.InteropServices")
+                {
+                    foreach (TypeDesc type in method.Instantiation)
+                    {
+                        AddPInvokeParameterDependencies(ref dependencies, factory, type);
+                    }
+                }
             }
-            else if (Internal.TypeSystem.Interop.MarshalHelpers.IsStructMarshallingRequired(parameter))
+        }
+
+        public static void AddPInvokeParameterDependencies(ref DependencyList dependencies, NodeFactory factory, TypeDesc parameter)
+        {
+            AddDependenciesDuePInvokeDelegate(ref dependencies, factory, parameter);
+            AddDependenciesDuePInvokeStruct(ref dependencies, factory, parameter);
+        }
+
+        public static void AddDependenciesDuePInvokeDelegate(ref DependencyList dependencies, NodeFactory factory, TypeDesc type)
+        {
+            if (type.IsDelegate)
+            {
+                dependencies.Add(factory.NecessaryTypeSymbol(type), "Delegate Marshalling Stub");
+
+                dependencies.Add(factory.MethodEntrypoint(factory.InteropStubManager.GetOpenStaticDelegateMarshallingStub(type)), "Delegate Marshalling Stub");
+                dependencies.Add(factory.MethodEntrypoint(factory.InteropStubManager.GetClosedDelegateMarshallingStub(type)), "Delegate Marshalling Stub");
+                dependencies.Add(factory.MethodEntrypoint(factory.InteropStubManager.GetForwardDelegateCreationStub(type)), "Delegate Marshalling Stub");
+            }
+        }
+
+        public static void AddDependenciesDuePInvokeStruct(ref DependencyList dependencies, NodeFactory factory, TypeDesc type)
+        {
+            if (MarshalHelpers.IsStructMarshallingRequired(type))
             {
-                var stub = (Internal.IL.Stubs.StructMarshallingThunk)factory.InteropStubManager.GetStructMarshallingManagedToNativeStub(parameter);
-                dependencies.Add(factory.ConstructedTypeSymbol(factory.InteropStubManager.GetStructMarshallingType(parameter)), "Struct Marshalling Type");
+                dependencies.Add(factory.NecessaryTypeSymbol(type), "Struct Marshalling Stub");
+
+                var stub = (Internal.IL.Stubs.StructMarshallingThunk)factory.InteropStubManager.GetStructMarshallingManagedToNativeStub(type);
+                dependencies.Add(factory.ConstructedTypeSymbol(factory.InteropStubManager.GetStructMarshallingType(type)), "Struct Marshalling Type");
```

Comment:
> we would need it in the marshaling stubs anyway (in locals, calli signature, etc).

Locals and calli signatures are exactly the kind of places that don't force an EEType because the runtime doesn't care. E.g. for locals the type information is dissolved by codegen into "how big is this type" (what's the size of it in the frame) and "what's the GC information" (this is part of the method's GC info). None of those refer back to the EEType.

We basically only need the EEType if someone boxes the struct, or has a typeof of it in code (or does an unbox, but that doesn't make sense without the former) - which is why I'm questioning forcing the EEType here. It's likely ~50 bytes of waste per marshallable struct.